File: FiveInRow.UI.Metro/Components/BoardPanel.cs
using FiveInRow.Core.UI;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;

namespace FiveInRow.UI.Metro.Components
{
    public class BoardPanel : Panel
    {
        #region .ctors

        public BoardPanel()
        {
            Loaded += BoardPanel_Loaded;
            RenderTransformOrigin = new Point(0.5, 0.5);
            RenderTransform = _tTransform;

            _eventProc = new EventProcImpl() { Panel = this, CellWidth = 60, CellHeight = 60 };
        }

        #endregion .ctors

        #region Fields

        private EventProc _eventProc;
        private bool _isPositioned;

        private TranslateTransform _tTransform = new TranslateTransform();
        private ItemsControl _owner;
        private MainPageViewModel _vm;

        #endregion Fields

        #region Methods

        public void Centrify()
        {
            _eventProc.Centrify();
        }

        void BoardPanel_Loaded(object sender, Windows.UI.Xaml.RoutedEventArgs e)
        {
            Loaded -= BoardPanel_Loaded;

            _owner = ItemsControl.GetItemsOwner(this);
            _vm = (MainPageViewModel)_owner.DataContext;
            _vm.SetPanel(this);
            _vm.SetOffset(new Point(_tTransform.X, _tTransform.Y));

            _owner.PointerPressed += (s, arg) =>
            {
                var pos = arg.GetCurrentPoint(_owner).Position;
                _eventProc.OnPointerPressed(arg.Pointer, _eventProc.CreatePos(pos.X, pos.Y));
            };
            _owner.PointerReleased += (s, arg) =>
            {
                var pos = arg.GetCurrentPoint(_owner).Position;
                _eventProc.OnPointerReleased(arg.Pointer, _eventProc.CreatePos(pos.X, pos.Y));
            };
            _owner.PointerMoved += (s, arg) =>
            {
                var pos = arg.GetCurrentPoint(_owner).Position;
                _eventProc.OnPointerMoved(arg.Pointer, _eventProc.CreatePos(pos.X, pos.Y));
            };
        }

        protected override Size MeasureOverride(Size availableSize)
        {
            if (Children.Count == 0) return base.MeasureOverride(availableSize);
            else
            {
                foreach (var item in Children)
                {
                    item.Measure(availableSize);
                }
                return availableSize;
            }
        }

        protected override Size ArrangeOverride(Size finalSize)
        {
            if (Children.Count > 0)
            {
                if (!_isPositioned)
                {
                    _isPositioned = true;
                    _eventProc.Centrify(finalSize.Width, finalSize.Height);
                }
            }

            foreach (var item in Children.OfType<FrameworkElement>())
            {
                var cell = (CellView)item.DataContext;
                item.Arrange(new Rect(new Point(_eventProc.CellWidth * (cell.Col - 1), _eventProc.CellHeight * (cell.Row - 1)), new Size(_eventProc.CellWidth, _eventProc.CellHeight)));
            }
            return finalSize;
        }

        #endregion Methods

        #region Internal classes

        private class EventProcImpl : EventProc
        {
            public BoardPanel Panel { get; set; }

            public override double Width { get { return Panel.ActualWidth; } }

            public override double Height { get { return Panel.ActualHeight; } }
            
            public override void CapturePointer(object obj)
            {
                Panel.CapturePointer((Pointer)obj);
            }

            public override void ReleasePointer(object obj)
            {
                Panel.ReleasePointerCapture((Pointer)obj);
            }

            public override void Set(int row, int col)
            {
                Panel._vm.Set(row, col);
            }

            public override void SetOffset(double dx, double dy)
            {
                Panel._tTransform.X = dx;
                Panel._tTransform.Y = dy;
                if (Panel._vm != null) Panel._vm.SetOffset(new Point(dx, dy));
            }
        }

        #endregion Internal classes
    }
}

File: FiveInRow.UI.Desktop/BoardPanel.fs
namespace FiveInRow.UI.Desktop

open System
open System.Collections.Generic
open System.Linq
open System.Text
open System.Windows.Controls
open System.Windows
open System.Windows.Input
open System.Windows.Media

open FiveInRow.Foundation

type BoardPanel() as this =
    inherit Panel()

    let mutable _cellWidth = 0.0
    let mutable _cellHeigth = 0.0

    let mutable _isCaptured = false
    let mutable _isDragging = false
    let mutable _lastPos = new Point(0.0, 0.0)
    let mutable _offset = new Vector(0.0, 0.0)
    let _tTransform = new TranslateTransform()

    let onLoaded e =
        let owner = ItemsControl.GetItemsOwner(this)
        let vm = owner.DataContext :?> MainWindowViewModel

        let onMouseDown (p: Point) =
            _isCaptured <- true
            _lastPos <- p
            ()

        let onMouseUp (p: Point) =
            _isCaptured <- false
            this.ReleaseMouseCapture()

            if _isDragging then _isDragging <- false
            else
                let c = p - _offset
                let index = (int(c.Y / _cellWidth) + 1, int(c.X / _cellHeigth) + 1)
                vm.Set index

        let onMouseMove(p: Point) =
            if _isCaptured then
                this.CaptureMouse() |> ignore
                let offset = p - _lastPos
                if _isDragging || offset.Length > 6.0 then
                    _isDragging <- true
                    _offset <- _offset + offset
                    _offset <- Vector(Math.Min(_offset.X, this.ActualWidth / 4.0), Math.Min(_offset.Y, this.ActualHeight / 4.0))
                    _tTransform.X <- _offset.X
                    _tTransform.Y <- _offset.Y
                    _lastPos <- p
                    vm.Offset <- _offset

        owner.MouseDown.Add (fun e -> owner |> e.GetPosition |> onMouseDown)
        owner.MouseUp.Add (fun e -> owner |> e.GetPosition |> onMouseUp)
        owner.MouseMove.Add (fun e -> owner |> e.GetPosition |> onMouseMove)
        ()

    do 
        this.Loaded.Add onLoaded
        this.RenderTransform <- _tTransform

    override x.MeasureOverride availableSize = 
        if x.InternalChildren.Count = 0 then base.MeasureOverride(availableSize)
        else
            for el in x.InternalChildren do
                el.Measure(availableSize)
            _cellWidth <- x.InternalChildren |> Seq.cast<FrameworkElement> |> Seq.map (fun t -> t.DesiredSize.Width) |> Seq.max
            _cellHeigth <- x.InternalChildren |> Seq.cast<FrameworkElement> |> Seq.map (fun t -> t.DesiredSize.Height) |> Seq.max
            availableSize

    override x.ArrangeOverride finalSize =
        for el in x.InternalChildren |> Seq.cast<FrameworkElement> do
            let cell = el.DataContext :?> CellView
            Rect(Point(_cellWidth * float(cell.Col - 1), _cellHeigth * float(cell.Row - 1)), Size(_cellWidth, _cellHeigth)) |> el.Arrange
        finalSize

